src/translator/cache.cpp

```diff
@@ -0,0 +1,157 @@
+#include "translator/cache.h"
+
+#include <cstdlib>
+
+namespace marian {
+namespace bergamot {
+
+#ifndef WASM_COMPATIBLE_SOURCE
+
+ThreadSafeL4Cache::ThreadSafeL4Cache(Ptr<Options> options)
+    : epochManagerConfig_(options->get<size_t>("cache-ebr-queue-size"),
+                          std::chrono::milliseconds(options->get<size_t>("cache-ebr-interval")),
+                          options->get<size_t>("cache-ebr-num-queues")),
+      cacheConfig_(options->get<size_t>("cache-size") * 1024 * 1024,
+                   std::chrono::seconds(options->get<size_t>("cache-time-to-live")),
+                   options->get<bool>("cache-remove-expired")),
+      service_(epochManagerConfig_),
+      context_(service_.GetContext()) {
+  // There is only a single cache we use. However, L4 construction API given by example demands we give it a string
+  // identifier.
+  const std::string cacheIdentifier = "global-cache";
+  size_t numBuckets = options->get<size_t>("cache-buckets");
+  hashTableIndex_ = service_.AddHashTable(L4::HashTableConfig(
+      cacheIdentifier, L4::HashTableConfig::Setting{static_cast<uint32_t>(numBuckets)}, cacheConfig_));
+}
+
+bool ThreadSafeL4Cache::fetch(const marian::Words &words, ProcessedRequestSentence &processedRequestSentence) {
+  auto &hashTable = context_[hashTableIndex_];
```

Comment:
Is there a reason for using a reference here as opposed to a pointer? Just a bit of an unusual design pattern.

Reply:
Another minor comment. Excessive use of `auto` makes readability worse.

Reply:
> Is there a reason for using a reference here as opposed to a pointer? 

Reason: L4 Example.

* https://github.com/microsoft/L4/blob/master/Examples/main.cpp#L9-L24

This is a remnant of copying over from the example and getting it to work. I can try to find the actual type and fix it if you insist, but I'd prefer to keep closer to the example and point there. Maybe adding a comment linking would help as an alternative?

Reply:
I'll fire up my editor over the weekend and propose some changes.

Reply:
auto here resolves to `IWritableHashTable` and gives a reference, so I guess you don't have much choice regarding the reference pattern here.